### PR TITLE
[tracking-transparency] Implemented `getAdvertisingId` for Android and iOS 

### PR DESCRIPTION
**Pull request by reichhartd**
# Why

At present, Expo's `TrackingTransparency` can only be employed to request tracking permissions. However, there are scenarios where we also require the [AAID](https://developer.android.com/training/articles/ad-id) for Android and the [IDFA](https://developer.apple.com/documentation/adsupport/asidentifiermanager/advertisingidentifier) for iOS.

# How

In response to this need, I've implemented the `getAdvertisingIdAsync` function in this pull request.
As this is my first major pull request for Expo, I appreciate any feedback 🙏 

I do have a few questions I'm uncertain about:
- Is the `getAdvertisingIdAsync` naming apt, or should we consider dividing it into `getAaidAsync`/`getAndroidAdvertisingIdAsync` for Android and `getIdfaAsync`/`getIosIdForAdvertisersAsync` for iOS?
- Should this package be renamed from `expo-tracking-transparency` to `expo-tracking` as it no longer exclusively deals with the transparency part?
- The [`getIosIdForVendorAsync`](https://docs.expo.dev/versions/latest/sdk/application/#applicationgetiosidforvendorasync) function is a part of `expo-application`. However, integrating `getAdvertisingIdAsync` there seems out of place, as this function holds little value without the tracking permission. Should all tracking methods live in this package?

# Test Plan

Tested it for Android in the emulator, and for iOS on my physical device.

# Checklist

- [X] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [X] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [X] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-tracking-transparency/ios/TrackingTransparencyModule.swift
Comment (tsapeta) [inline packages/expo-tracking-transparency/ios/TrackingTransparencyModule.swift:31]: Looks like on both iOS and Android the returned UUID can be with all zeros. I'm wondering if it wouldn't be better to just return `null` in that case. The user will usually have to do one more check whether the id is present, but it increases the awareness that this value may be unavailable in some scenarios. What are your thoughts?

File: packages/expo-tracking-transparency/src/TrackingTransparency.ts
Comment (tsapeta) [inline packages/expo-tracking-transparency/src/TrackingTransparency.ts:119]: If the native function has a return type, it will never return or resolve to `undefined` - it would be `null` instead.

```suggestion
export async function getAdvertisingIdAsync(): Promise<string | null> {
```

**Review by tsapeta — CHANGES_REQUESTED** [review]
Thanks and congrats for your first PR! 👏 

> Is the getAdvertisingIdAsync naming apt, or should we consider dividing it into `getAaidAsync`/`getAndroidAdvertisingIdAsync` for Android and `getIdfaAsync`/`getIosIdForAdvertisersAsync` for iOS?

Yeah, I think `getAdvertisingId` (without `Async`) is good. This identifier means basically the same on both platforms, so it makes sense to have just one function.

> Should this package be renamed from `expo-tracking-transparency` to `expo-tracking` as it no longer exclusively deals with the transparency part?

I agree that `expo-tracking` might be better now, but I would rather keep using the current name so we don't have to introduce a breaking change. At some point we may want to move all SDK packages to `@expo` scope on npm and then we'll need a script that migrates all the imports during SDK upgrade – that will be a good occasion to rename this package.

> The [`getIosIdForVendorAsync`](https://docs.expo.dev/versions/latest/sdk/application/#applicationgetiosidforvendorasync) function is a part of `expo-application`. However, integrating `getAdvertisingIdAsync` there seems out of place, as this function holds little value without the tracking permission. Should all tracking methods live in this package?

You're right, let's keep it in `expo-tracking-transparency`. To be honest I was even thinking about deprecating/dropping `expo-application` 😅

File: packages/expo-tracking-transparency/src/ExpoTrackingTransparency.ts
Comment (reichhartd) [reply packages/expo-tracking-transparency/src/ExpoTrackingTransparency.ts:3]: You're welcome 🙏

File: packages/expo-tracking-transparency/ios/TrackingTransparencyModule.swift
Comment (reichhartd) [reply packages/expo-tracking-transparency/ios/TrackingTransparencyModule.swift:31]: Good thought 🙏 
Since it's the same for iOS and Android, I'll do the check on the JS side.

**Comment by ide** [comment]
One thing to watch out for is static analysis from app store reviews. I don't know the latest but it is possible that apps that contain code that could access the IDFA are subject to greater review and may need to explain how they use the IDFA, which would be a problem for apps that use expo-tracking-transparency and don't need the IDFA.

**Comment by reichhartd** [comment]
@tsapeta Ah I already thought `@expo/video` was special because `expo-video` was already claimed 😄.
In response to your feedback, I've addressed all comments, enhanced the documentation, revised the test cases, and added `TODO` comments for a future renaming of this package to `ExpoTracking`. I've also migrated `getAndroidId` and `getIosIdForVendorAsync` from `expo-application` to this package.

@ide I searched online about your concern but didn't find anything concrete. If preferred, `getAndroidId` and `getIosIdForVendorAsync` can certainly remain within the `expo-application` package.

If you could review it again, that would be great. I would love to get this into SDK 50 🙏

File: packages/expo-tracking-transparency/src/TrackingTransparency.ts
Comment (tsapeta) [inline packages/expo-tracking-transparency/src/TrackingTransparency.ts:47]:
```suggestion
  if (advertisingId === '00000000-0000-0000-0000-000000000000') {
    return null;
  }
```

File: packages/expo-tracking-transparency/android/src/main/java/expo/modules/trackingtransparency/TrackingTransparencyModule.kt
Comment (tsapeta) [inline packages/expo-tracking-transparency/android/src/main/java/expo/modules/trackingtransparency/TrackingTransparencyModule.kt:21]: As I commented above – let's keep it only in `expo-application` for now.

```suggestion
```

**Review by tsapeta — CHANGES_REQUESTED** [review]
Thanks for applying these changes! Generally it looks good to me, just please revert changes related to `expo-application` 😉

**Comment by ide** [comment]
> @ide I searched online about your concern but didn't find anything concrete. 

Thanks for searching. My main concern was about the IDFA since it is perceived to be more of a liability than an asset now and we want developers who aren't using it not to be negatively impacted by it, and in recent years app store reviews have sometimes asked for unused/dead code to be removed if it statically accesses sensitive APIs.

If this isn't an actual problem, it's fine for now. If app stores start rejecting apps due to unused IDFA-related code, we'll need to remove these APIs.

**Comment by reichhartd** [comment]
@tsapeta I adopted your requested changes.

@ide I get your point. My point is, that if users have installed the `expo-tracking-transparency` package, it's likely that they are already making an IDFA call (for example, via another third-party library, such as `react-native-appsflyer` or `@segment/analytics-react-native`). In addition, I wonder if Apple analyses the JS bundle if the function `getAdvertisingId` in `expo-tracking-transparency` is called on Swift side. 

If this becomes a problem, I suggest moving these APIs to an extra package instead of removing them.

**Review by tsapeta — APPROVED** [review]
Looks good to me now!

> If this becomes a problem, I suggest moving these APIs to an extra package instead of removing them.

One more thing that we can do before removing/moving is to cheat the static analyzer by creating a selector from concatenated string and then performing it, see https://github.com/expo/expo/blob/ab01a2286925b63fb01056e168b790b23a3be622/packages/expo-location/ios/EXLocation/Requesters/EXBackgroundLocationPermissionRequester.m#L33

Afaik many 3rd-party libraries do that, sometimes even in a more sophisticated way.

I'll wait for @ide's approval before merging.

**Comment by ide** [comment]
> cheat the static analyzer

Other libraries may choose to do that but Expo's code should follow the intent of app review.

File: packages/expo-tracking-transparency/src/TrackingTransparency.ts
Comment (ide) [inline packages/expo-tracking-transparency/src/TrackingTransparency.ts:44]:
```suggestion
/**
 * Gets the advertising ID, a UUID string intended only for advertising. Use this string for
 * frequency capping, attribution, conversion events, estimating the number of unique users,
 * advertising fraud detection, and debugging.
 *
 * As a best practice, don't store the advertising ID. Instead, call this function each time your
 * app needs to use the advertising ID. Users can change whether they allow app tracking and can
 * reset their advertising ID at any time in their system settings. Check your app's authorization
 * using `getTrackingPermissionsAsync()` to determine the user's intent.
 *
 * On Android, this function returns the "Android Advertising ID"
 * ([AAID](https://developers.google.com/android/reference/com/google/android/gms/ads/identifier/AdvertisingIdClient.Info#public-string-getid)).
 * On Android devices that support multiple users, including guest users, it's possible for your app
 * to obtain different advertising IDs on the same device. These different IDs correspond to
 * different users who could be signed in on that device. See Google's documentation for more
 * information: [Get a user-resettable advertising
 * ID](https://developer.android.com/training/articles/ad-id).
 *
 * On iOS, this function returns the "Identifier for Advertisers"
 * ([IDFA](https://developer.apple.com/documentation/adsupport/asidentifiermanager/advertisingidentifier)),
 * a string that's unique to each device. On devices running iOS 14.5 and newer, your app must
 * request tracking authorization using `requestTrackingPermissionsAsync()` before it can get the
 * advertising identifier.
 *
 * @return Returns either a UUID `string` or `null`. It returns null in the following cases:
 * - On Android, when `isLimitAdTrackingEnabled()` is `true`
 * - In the iOS simulator, regardless of any settings
 * - On devices running iOS 14.5 and later if you haven't received permission using
 *   `requestTrackingPermissionsAsync()`
 * - On iOS, if you've requested permission and the user declines
 * - On iOS, when a profile or configuration restricts access to the advertising identifier, such as
 *   when the user has turned off the system-wide "Allow Apps to Request to Track" setting
 *
 * @example
 * ```ts
 * TrackingTransparency.getAdvertisingId();
 * // "E9228286-4C4E-4789-9D95-15827DCB291B"
 * ```
 */
```

File: packages/expo-tracking-transparency/android/src/main/java/expo/modules/trackingtransparency/TrackingTransparencyModule.kt
Comment (ide) [inline packages/expo-tracking-transparency/android/src/main/java/expo/modules/trackingtransparency/TrackingTransparencyModule.kt:23]: @tsapeta can you please confirm this will get converted to null in JS (not undefined)?

**Review by ide — APPROVED** [review]

**Comment by ide** [comment]
> My point is, that if users have installed the expo-tracking-transparency package, it's likely that they are already making an IDFA call

I agree with this.

**Comment by gabrieldonadel** [comment]
Was this tested using Expo Go? I'm getting the following error 
image

**Comment by reichhartd** [comment]
No, unfortunately not. What are the necessary steps for enabling Expo Go support for the new Android package? Should it be included in the [`getModulesList`](https://github.com/expo/expo/blob/main/android/expoview/src/main/java/versioned/host/exp/exponent/ExperiencePackagePicker.kt#L119)? Additionally, there is a Gradle dependency specified [here](https://github.com/expo/expo/blob/main/packages/expo-tracking-transparency/android/build.gradle#L107). Does this dependency also need to be integrated?

**Pull request merged**
